ui: Navigation panel redesign

koorosh opened the pull request:
Related task: #42920
Depends on PR: #43181

`SideNavigation` panel on the left side displays the full text of links so it is easier to suggest what it means.
With this change, following is affected:
- Layout of the main panel is changed to adjust to the new width
- `GlobalNavigation` and `TabNavigatin` panels are added on top of the screen as fixed panels that aren't scrolled out. A scroll bar appears for the main panel only when it's content doesn't fit on a screen.

`TabNavigation` panel displays:
- cluster name (in case it is defined) otherwise cluster id;
- Cluster version in a badge;

Screens:
<img width="723" alt="Screenshot 2019-12-17 at 13 53 55" src="https://user-images.githubusercontent.com/3106437/70994432-e7aa2880-20d6-11ea-911a-0cf2a88c41bf.png">

<img width="723" alt="Screenshot 2019-12-17 at 13 54 28" src="https://user-images.githubusercontent.com/3106437/70994433-e842bf00-20d6-11ea-8a31-f24d752fc862.png">

<img width="723" alt="Screenshot 2019-12-17 at 13 54 04" src="https://user-images.githubusercontent.com/3106437/70994435-e842bf00-20d6-11ea-88fc-f7f4a822105a.png">

koorosh committed:
ui: Refactor LoginIndicator to standalone component

Release note: None

koorosh committed:
ui: Redesign Navigation panel view and Main Navigation panel

Release note: None

koorosh committed:
ui: Fix imports for design system components

Release note: None

koorosh committed:
ui: Display cluster name and version on tab navigation header

- Retrieve cluster name and version from the first live node in a cluster

Release note: None

koorosh committed:
ui: Select clusterVersion, clusterId, and clusterName on `layout` container

- Move clusterVersion selector to `nodes` reducer as this value is retrieved from nodes description;
- Select clusterVersion, clusterId, and clusterName on `layout` container.

Release note: None

koorosh committed:
ui: Add 'Cluster id' description as a description on TabNavigation panel

It is not clear what displayed UUID means on TabNavigation panel without any additional
description. To fix this, 'Cluster id:' prefix is added as an explicit description.

Release note: None

koorosh committed:
ui: Fix imports for 'versionsSelector' function in tests

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/43236)
<!-- Reviewable:end -->

dhartunian reviewed (changes requested):
> Reviewed 2 of 35 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43236)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Koorosh)

___
*[pkg/ui/src/components/tabNavigation/tabNavigation.tsx, line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43236#-Ly5uhP6EmtcfIlkDhcc:-Ly5uhP6EmtcfIlkDhcd:b-2rn8b) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbb03c513d105009478e17c2a86ff9e6584b6a91/pkg/ui/src/components/tabNavigation/tabNavigation.tsx#L15)):*
> ```XML
> import "./tabNavigation.styl";
> 
> export interface TabNavigationProps {
> ```

Why is this called `TabNavigation` instead of something like `Header`? It doesn't seem like there's any interaction portion. Or is that coming up?
___
*[pkg/ui/src/components/text/text.tsx, line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43236#-Ly5utXv2ug1-tNyB0_B:-Ly5utXv2ug1-tNyB0_C:b92t21i) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbb03c513d105009478e17c2a86ff9e6584b6a91/pkg/ui/src/components/text/text.tsx#L77)):*
> ```XML
>   );
>   return (
>     <span className={textTypeClass}>
> ```

Can we somehow use `h1`, `h2`, etc. tags here instead of `span` for everything?
___
*[pkg/ui/src/views/app/components/layoutSidebar/index.tsx, line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43236#-Ly62unp6222JzJQowAm:-Ly62unp6222JzJQowAn:b-z0crj1) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbb03c513d105009478e17c2a86ff9e6584b6a91/pkg/ui/src/views/app/components/layoutSidebar/index.tsx#L35)):*
> ```XML
> 
>   isActiveNavigationItem = (path: string): boolean => {
>     const { activeFor } = this.routes.find(route => route.path === path);
> ```

I think you can pass in `activeFor` into the arrow function below and then into this function so no need to re-construct it. 


<!-- Sent from Reviewable.io -->

dhartunian commented:
@Koorosh can you take a quick look at Firefox for layout issues too? I noticed the new panels along the top getting squashed when you clicked on different links in the left panel

<img width="1792" alt="Screen Shot 2020-01-08 at 4 55 15 PM" src="https://user-images.githubusercontent.com/986307/72019490-d072dc80-3237-11ea-850f-6f9d8112dc84.png">

koorosh committed:
ui: Rename TabNavigation component to PageHeader

Release note: None

koorosh committed:
ui: Refactor Text component to use Heading tag elements

This change required to make changes for globally applied styles for
heading elements to make changes in isolation.
- Existing global styles are refactored to `h.base-heading` class.
- Text component builds tag element conditionally based on input TextType
property. Instead of using JSX constructor, React.createElement is used,
so it is possible to construct element with dynamic string literal which
represents desired element name.

Release note: None

koorosh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43236)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @Koorosh)

___
*[pkg/ui/src/components/text/text.tsx, line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43236#-Ly5utXv2ug1-tNyB0_B:-LyEpxox-s1hiIt1s60O:bebrt1w) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbb03c513d105009478e17c2a86ff9e6584b6a91/pkg/ui/src/components/text/text.tsx#L77)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Can we somehow use `h1`, `h2`, etc. tags here instead of `span` for everything?
</blockquote></details>

Will do
___
*[pkg/ui/src/components/tabNavigation/tabNavigation.tsx, line 15 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43236#-Ly5uhP6EmtcfIlkDhcc:-LyA8TBV6O1uMWaLSU0K:b-tw2cdo) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbb03c513d105009478e17c2a86ff9e6584b6a91/pkg/ui/src/components/tabNavigation/tabNavigation.tsx#L15)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Why is this called `TabNavigation` instead of something like `Header`? It doesn't seem like there's any interaction portion. Or is that coming up?
</blockquote></details>

The names have been borrowed :) from Design System component descriptions in Zeplin (https://app.zeplin.io/project/5c3bf1d79ebc14be7741eae6/screen/5db87b1f42347622db611247)
According to designs it might or might not have a breadcrumb or tab items inside.
I agree, for now it doesn't look like a perfect name. Will be changed and then the `TabNavigation` component can be developed as another separate component.


<!-- Sent from Reviewable.io -->

koorosh committed:
ui: Adjust header's components with a small screen sizes

When screen width becomes smaller, Header's components wrap their
content and it looks very odd. To avoid this, Header component wraps
their content so it is stacked.

Release note: None

koorosh committed:
ui: Fix top navigation panel height sizing (Firefox compatibility issue)

It looks like Firefox calculate flex-basis height or height of flex-items
in a different way than Chrome browser. To avoid current issue, min-height
ensures that height of the component won't change.

Release note: None

koorosh committed:
ui: Add browserslist section to package.json

It is initial default configuration to track which browsers versions
have to be supported.
Default configuration includes: (> 0.5%, last 2 versions, Firefox ESR,
not dead).

Release note: None

Koorosh commented:
Current branch has been rebased to resolve conflicts with current state of `master` branch

koorosh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43236)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @Koorosh)

___
*[pkg/ui/src/views/app/components/layoutSidebar/index.tsx, line 35 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43236#-Ly62unp6222JzJQowAm:-LyTN9TWAp46L4XVaBEo:bo4txcb) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbb03c513d105009478e17c2a86ff9e6584b6a91/pkg/ui/src/views/app/components/layoutSidebar/index.tsx#L35)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

I think you can pass in `activeFor` into the arrow function below and then into this function so no need to re-construct it. 
</blockquote></details>

- I would say that `render` method is not responsible for providing `activeFor` argument because it is accessible from every place inside of class instance;
- `this.isActiveNavigationItem` method has to decide whether provided `path` is active or not and entire logic (which depends on `activeFor` parameter has to be hidden from method consumer;


<!-- Sent from Reviewable.io -->

koorosh reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43236)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian)

___
*[pkg/ui/src/components/text/text.tsx, line 77 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/43236#-Ly5utXv2ug1-tNyB0_B:-LyTQRKB69NXpJABsBHR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dbb03c513d105009478e17c2a86ff9e6584b6a91/pkg/ui/src/components/text/text.tsx#L77)):*
<details><summary><i>Previously, Koorosh (Andrii Vorobiov) wrote…</i></summary><blockquote>

Will do
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 

> Reviewed 16 of 35 files at r1, 2 of 5 files at r3, 3 of 5 files at r4, 2 of 3 files at r6, 2 of 2 files at r8, 3 of 5 files at r9, 39 of 41 files at r10, 6 of 7 files at r11, 1 of 1 files at r12, 1 of 1 files at r13.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/43236#-:-Lz9Q8Mu5I2_-MiKImSx:bg7igkv)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dhartunian)



<!-- Sent from Reviewable.io -->

dhartunian commented:
@Koorosh can you rebase and resolve conflicts? I will then merge this tomorrow.

Koorosh commented:
> @Koorosh can you rebase and resolve conflicts? I will then merge this tomorrow.

@dhartunian , done.

dhartunian commented:
bors r+
